Refactor forecast

isaacaka opened the pull request:
The aim is to refactor the forecast.py code to make it easier to swap between different dimensionality reduction and forecasting techniques.

- Refactored `jumper.ipynb`, `main_forecast.py` and main to reflect changes to the refactored libraries. The jumper notebook is specific to using normal PCA (not Kernal PCA)
- Separated code into forecasting technique classes and dimensionality reduction classes
- `forecast.py` calls the methods defined in the above two classes
- The DR and forecasting techniques created are imported and stored in a dictionary
- The DR and forecasting techniques the user wants to use can be specified by setting the name within `techniques_config.yaml`

- `ocean_terms.yaml` defines the names of the terms stored in the NEMO grid files, e.g. different configurations of nemo use different names for temperature, ssh and salinity. You can specify which names are used in the yaml file. and `restart.py` will then be able to read the correct keys.

Tested that the refactored code runs end-to-end

Test written against the non-refactored code need to be written. 

Closes #37 #52 #9 #44 #48 #58 #4

isaacaka committed:
Refactor forecasting pipeline; add Gaussian process & fixes

- Move component reconstruction to Simulation class, functionalise getting
number of components, predict with skforecast
- Refactor forecasting code to implement an abstract base
- Add non-skforecast based gaussian process
- Fix typo in name of forecaster class
- Add skforecast to requirements.txt
- Change rmse function to more general error function in ABC class
- Add assert on the required dimensionality of the data
- Change logic for how number of years is forecasted to fix issue where,
number of years forecasted were incorrect
- Change to append whole map instead of first value in map
- Call the correct names of files for forecasted components

isaacaka committed:
Refactor technique configuration; update parameters and checks

- Change parameters from tuple to seperate named variables
- Reorganise specifying DR and forecasting techniques
- Add a config file for specifying DR/forecaster techniques
- Add config file for specifying techniques
- Fix import error
- Add config file for names of forecasted ocean terms
- Add additional dimensionality check
- Change config file structure
- Change naming from rmse to error
- Remove TODO comments

isaacaka committed:
Refactor DR and forecasting setup; update configs and paths

- Create a setter method to pass neccessary variables from simulation
class to the decompostion class
- Correct method signature for get_component in forecast.py and
dimensionality_reduction.py
- Change Forecaster parameters, adds placeholder values for standard
deviation
- Adds filepaths
- Rename config.yaml -> techniques_config.yaml

isaacaka committed:
Update notebook to reflect refactor, parameterise forecaster

- Update notebook to reflect refactor of libraries
- Parameterise classes, forecaster can now take a regressor as input
- Format notebook

isaacaka committed:
Refactor forecasting techniques; clean params, naming, and checks

- Parameterises create_gp_regressor
- Remove unused args
- Remove conditional handling of predict arguments
- Raise an error when an invalid technique is specified
- Fix typo in jumpper notebook, rename dictionary key in forecast.py
- Change the method call from getPC to get_component
- The getPC method was refactored to be the get_component
method
- Change name of dimensionality reduction technique
- Asserts on the shape within the dimesionality reduction code has been
removed as this is tested in seperate unit tests.

ma595 commented on Notebooks/Jumper.ipynb:
A few initial comments.

ma595 replied:
Make sure that `ruff check <file.py` passes. We need to add this as a check in the CI as well. I can add the linting rules I've used for `Spinup-Evaluation` as a new PR.

ma595 reviewed (changes requested):
Running through Jumper.ipynb. Some small nitpicking issues (will look over main structural changes next).

I deployed with python==3.11 and ran the Jumper.ipynb:
- I modified the path to point to `Grid_T` files.  
- Error is returned with KPCA due to the setting of 0.9 for components
- I modified the `DR_technique` in `techniques_config.yaml` to "`DimensionalityReductionPCA`"
  - Could we alias these names to make them a bit more readable?
  - I modified "`DimensionalityReductionPCA`" to "`DimensionalityReductionABC`" and I get a `KeyError 'DimensionalityReductionABC'` - could a more helpful error message be provided?
- Word "Compilated" is not correct (remnant of Maud's notebook).
- Error bars look a little weird (using jupyter notebook defaults) - I loaded the data in the google drive. 
<img width="607" alt="image" src="https://github.com/user-attachments/assets/2a398fb8-f1f6-495b-b705-742ea6ff49e1" />

- `simu_predicted` needs to be manually created otherwise predictions will fail (again a remnant from Maud's code). Consider adding to cell 12 - where it was previously. 
- When plotting evaluation `AttributeError: 'Simulation' object has no attribute 'getPC'` - does the `getPC` function exist anymore? Perhaps this isn't important?

ma595 commented on Notebooks/Jumper.ipynb:
I think I reviewed this previously. Will run through again and check if it runs without error.

ma595 commented on lib/dimensionality_reduction.py:
Nicely modularised now. `getPC()` -> `get_component()`.

General comments
- Need some tests
  - Avoid asserts - replace with ValueError or whatever's more relevant. 
  - Probably better as a PR

isaacaka replied:
Yes

ma595 commented on lib/forecast.py:
```diff
@@ -652,18 +673,12 @@ def reconstruct(self, predictions, n, begin=0):
+        # TODO: Change to DR_technique
+        self.int_mask, ts_array = DimensionalityReductionPCA.reconstruct_predictions(
+            predictions, n, self.info, begin
+        )
```
Would be great to generalise this now to `DimensionalityReductionMethod` or similar

isaacaka replied:
Yes

ma595 commented on lib/forecast.py:
```diff
@@ -547,7 +566,9 @@ def prepare(self, n, train_len, steps):
         """
         x_train = np.linspace(0, 1, train_len).reshape(-1, 1)
         pas = x_train[1, 0] - x_train[0, 0]
-        x_pred = np.arange(0, (len(self) + steps) * pas, pas).reshape(-1, 1)
+        # x_pred = np.arange(0, (len(self) + steps) * pas, pas).reshape(-1, 1) #TODO: Check this len(self) + steps logic
```
Has this defintely been checked? I recall discussing this - but I can't remember what we decided on. I suspect the old implementation was correct?

isaacaka replied:
It was correct bar an off-by-one error but needs to be generalised for arbitrary forecasting techniques.

ma595 commented on lib/forecast.py:
```diff
@@ -20,9 +22,53 @@
+    )
+else:
+    Forecast_technique = Forecast_techniques[config["Forecast_technique"]["name"]]
+
```
Would it be best to move all of this to somewhere else like a `main_forecast.py`?  See file comments.

isaacaka replied:
Yes

ma595 commented on lib/forecast.py:
- STYLE: Lower case `Forecast` function
- STYLE: There's a `prepare` method in `Simulation` class as well as `main_forecast` and `Predictions` class. It's quite confusing (I realise this is a remnant of Maud's code)

ma595 replied:
There are similarly named functions in `Simulation` and `Prediction` too, which makes the whole code quite hard to navigate.

ma595 commented on lib/forecast_technique.py:
This defines the forecasting strategy i.e. recursive, direct and the regressor i.e. GP or regression. 

- Regressor is defined here. 
	- GP, regression etc.
- Do all regressors have `predict(return_std=True)`?
	- suggests that all regressors have uncertainty outputs
- Lower case these i.e.:  `GaussianProcessRecursiveForecaster` to `gp_recursive_forecaster`?
	- They are instances of a class (i.e a Forecaster that takes a regressor as parameter).
	- But they are treated in much the same way as dimensionality reduction classes - this is potentially confusing. 
- Rename `ForcastTechnique` to something like `ForecastMethod` or `ForecastStrategy`.
- Could this code be generalised further? 
	- Drawbacks: Hardcoded forecast steps / lags in GP and quite tight coupling to skforecast
	- I could imagine a situation where instead of hardcoding combinations like you have done. You could 'register' different regressors and 'forecasters'. And then use a function to combine them. 
	- And use a configuration yaml file which specifies the regressor and forecast method you want a long with their parameters - so they aren't hardcoded. 
	- We can leave this to a future PR if you want. 
	- Happy to help with a PR here or provide an example. 
	- OR we could just define all of the available options a little more cleanly - this is better if we only expect to have a few regressor + forecaster combos.
- We could move instantiation of regressors and forecasters into a separate module.
- Testing

isaacaka replied:
## Purpose

Provide a library‑agnostic hook for any forecasting approach.

Only requirement: inherit from BaseForecaster and implement apply_forecast.

This would be the most general way for users to swap out different methods.


## Interface contract

Signature:
```
def apply_forecast(self, y_train, x_train, x_pred) -> y_pred
```
Inputs: your choice of arrays/Series/features.

Output: predictions aligned to x_pred (optional: also return uncertainty).

## We don’t enforce

No forced pairing of regressors + forecaster wrappers, no enforced hard‑coded steps or library imports.

I implemented an example sklearn+skforecast version to illustrate usage, but it’s just one of many ways the forecasting step could be done.

This approach allows for maximum customisability/flexibility for the user

ma595 commented on lib/restart.py:
- This is nicely abstracted now with the `get_ocean_term` function. The idea, I suppose, is to generalise this "toolbox" to work for other restart files i.e. Oceananigans? 
	- This `get_ocean_term()` function needs a docstring. I understand it gets the name of the generated `[so,toce,soce].npy` file. 
- STYLE: Think about the naming of the variables in this file. I would avoid thetao, so and zos / ssh. The restart convention of NEMO / DINO is heavily embedded in here. (`restart["tn"]`, `restart["sshn"]` etc.)

isaacaka replied:
I agree the names should be less NEMO/DINO specific. I've pretty much kept the naming the same as we received the codebase. I think this is a case where we would need to consult with Julie and co as they have a better understanding of the variable name meanings and how to best name similar properties etc. 

Thing like this could then be pulled out to a config file: `restart["sshn"]`

ma595 replied:
OK, thanks for this. Probably best to add this as an issue and then we can resolve this conversation.

ma595 replied:
@isaacaka can you confirm this has been done?

isaacaka replied:
Yes, done

ma595 commented on main_forecast.py:
-  Would be nice to have the `config` file that you're hinting at in `main_forecast.py`. I.e. abstract out the DINO and ssh/soce/toce mapping. You are already using this in the `restart.py` - could it also be used here as well?

isaacaka replied:
Yes, missed that TODO, when refactoring restart.py

ma595 reviewed (changes requested):
This PR introduces skforecast Forecasters, sklearn regressors and a dimensionality reduction class. 

Within forecast_technique.py we create instances of the combined regressor and forecaster. So we essentially create all available options here. 

These are then imported in the forecast.py

At present, all forecasters, regressors and dimensionality reduction algorithms are imported to the user in the `forecast.py`

Added new files:
- `dimensionality_reduction.py`
- `forecast_technique.py`

Configuration files
- `ocean_terms.yaml`
- `techniques_config.yaml`

`ocean_terms.yaml` now generalises the inputs quite nicely. Used in `restart.py`. 

## General Issues

- [x] - **Currently in a separate branch:** Add a quick sketch of an updated `README.md` (already mentioned). Generally we need more comments and docs. 
- [x] - **Currently in a separate branch:** Testing - We need this project wide. I'll help with this. 
- [x]   - **Currently in a separate branch:** Testing. Avoid assertions in favour of Exceptions. 
- [x] - Formatting / linting issues need addressing - To fix some automatically: Run `ruff check --fix`  . Others will need manually fixed. Make sure local install of Ruff is up to date `ruff check` 
- [x] - STYLE: Inconsistent formatting in docstrings. Consider adopting `numpy` style https://numpydoc.readthedocs.io/en/latest/format.html#parameters. I've added a PR that imposes this style project wide using github actions. `create_gp_regressor` in `lib/forecast_technique.py` is a good example of how to do it. 
- [ ] - **(Can be done in a seperate PR)** STYLE: Could you add some static type hinting where you think it'd be useful? This is to get us ready for `mypy` in future. 
- [x] - STYLE: Consider renaming `forecast_technique.py` to `forecast_method.py`? 
- [ ] - FUNCTIONALITY: I  don't think the parallel jobs in Forecast works.  (MINOR) **_(Number of jobs is set to 1 by default and currently works like this. Can review what happens when jobs > 1 and fix any issues if present in a separate PR)_**
- [x] - STYLE: `snake_case` for function and method names - class names have CamelCase / PascalCase. Again, these are remnants from Maud's code. 
- [x] - Remove optimisation from `lib/forecast.py` move to a different class? I'm not sure what this is currently used for? 
- **Detailed correct usage of forecast_technique: Regressor and Forecaster code is just an example, code allows the user full customisability** Possibly need to refactor forecast_technique again (and rename). 
- **Can be done in another PR:** Could we add a simple regressor like linear or tree based?
- **Detailed correct usage of forecast_technique: Regressor and Forecaster code is just an example, code allows the user full customisability** We need to think about how to generalise this further so we do not hardcode the regressor and Forecaster?

isaacaka committed:
Refactor forecasting structure; move configs, utils, and clean code

- Generalise call on DimensionalityReductionPCA to ensure the
reconstruct_predictions call is made on a variablewhich contains
the specified DR technique
- Move dict[classname, class] from forecast.py to forecast_technique.py
and dimensionality_reduction.py respectively
- Move reading of config file which contains the forecasting and
dimensionality reduction techniques used, to main_forecast. Passed
these classes/class instances to the Prediction and Simulation
classes
- Move get_ocean_term function to a utils.py file and add keyError
exception, import get_ocean_term function into restart.py
- Use the get_ocean_term function in main_forecast to get property
name corresponding from ocean_terms.yaml
- Rename forecast_technique.py to forecast_mathod.py
- Remove optimization class from forecast.py
- The optimization class is not used within the codebase. It seems
it may have been used to test different kernels for the GP.
- Remove unused imports
- Convert documentation to numpy style and to snake case

isaacaka committed:
Refactor forecasting workflow; add utils and rename methods

- Remove reconstruct method Prediction class as the reconstruct method is
duplicated across both the Simulation and Prediction class
- Create a utils library for helper functions
- Move the get_ocean_terms functions to utils
- Add the functions get_dr_technique and get_forecast_technique which
reads which technique to use from the config file.
- Rename functions to be more specific, fix plot interval
- Rename Simulation.prepare method to get_simulation_data
- Rename Predictions.Forecast method to parallel_forecast
- Rename Predictions.forecast_ts to forecast_single_series
- Rename Predictions.prepare to train_test_series
- Change x-axis plot numbers to reflect the change to only predicting on the test data/interval
- Use created functions to get instances of techniques used
- Update notebook to reflect refactor of codebase

isaacaka committed:
Clean codebase and add README for custom technique integration

- Remove unused imports and variables
- Exclude unused and restart related code
- Update docstrings
- Add a README to show how a user can implement custom techniques
- The README contains examples of how to impplement custom forecasting
and dimensionality reduction techniques, how to specify which techniques
to use via config files. Examples of using ocean_terms.yaml as a
dictionary to match variables with their model specific reference
e.g Salinity : soce
- Update structure of README and add instruction for adding custom
techniques
- Merge refactor_README into README
- Remove refactor_README

isaacaka committed:
Refactor tests and fixtures for forecast refactor; update paths

- Refactor tests and fixtures to match updated Predictions class and
forecast_single_series method
- Fix indexing issue in forecast_single_series uncovered during
testing
- Update test names and logic to reflect forecasting only on unseen
interval.
- Rename functions
(prepare → get_simulation_data, applyPCA → decompose, etc.) for clarity.
- Update Simulation fixture and imports following refactor.
- Adjust handling of term variable (now a string instead of a tuple).
- Remove redundant tests using random data; retained those using real
data.
- Set time_dim key correctly, resolving test failures.
- Move prepare function to utils and fixed related imports.
- Modify output directory creation to generate unique, timestamped paths
per run.

isaacaka committed:
Updated README, fix linting issues

isaacaka committed:
Add pytest to requirements.txt

ma595 reviewed (dismissed):
Thanks for making the changes to the README.md. LGTM.

ma595 reviewed (approved):
Checked commit history and README. LGTM
